File: desktop/src/main/diagnostics.ts
// The "Copy diagnostics" payload: everything worth pasting into a chat when
// something looks wrong, and nothing that shouldn't leave the machine. The
// token, the log's own scrubbing, and any URL credentials are excluded by
// construction — this text is meant to be shared.

import { app, Notification, powerMonitor } from 'electron'
import os from 'node:os'
import { logPath, logText } from './log'
import type { Settings } from './settings'
import type { UpdateState } from './updater'

export interface DiagnosticsInput {
  settings: Settings
  update: UpdateState
  watchConnected: boolean
  appdVersion: string | null
  lastError: string | null
}

export function buildDiagnostics(input: DiagnosticsInput): string {
  const s = input.settings.view()
  const mem = process.memoryUsage()
  const lines = [
    `# Huginn Desktop diagnostics`,
    `generated       ${new Date().toISOString()}`,
    ``,
    `## App`,
    `version         ${app.getVersion()}`,
    `packaged        ${app.isPackaged}`,
    `electron        ${process.versions['electron'] ?? '?'}`,
    `chrome          ${process.versions['chrome'] ?? '?'}`,
    `platform        ${process.platform} ${process.arch} (${os.release()})`,
    `uptime          ${Math.round(process.uptime())}s`,
    `heap            ${Math.round(mem.heapUsed / 1024 / 1024)}MB used / ${Math.round(mem.rss / 1024 / 1024)}MB rss`,
    `idle            ${powerMonitor.getSystemIdleTime()}s`,
    ``,
    `## Connection`,
    `server          ${s.baseUrl}`,
    `token           ${s.hasToken ? 'set' : 'MISSING'}${s.tokenPlaintextFallback ? ' (plaintext fallback — no OS keyring)' : ''}`,
    `install id      ${s.installId}`,
    `watch stream    ${input.watchConnected ? 'connected' : 'DISCONNECTED'}`,
    `appd version    ${input.appdVersion ?? 'unknown'}`,
    ``,
    `## Notifications`,
    `enabled         ${s.notifyEnabled}`,
    `os support      ${Notification.isSupported()}`,
    `claiming route  ${s.notifyEnabled && Notification.isSupported() && powerMonitor.getSystemIdleTime() < 600}`,
    ``,
    `## Update`,
    `status          ${input.update.status}`,
    `version         ${input.update.version ?? '-'}`,
    `error           ${input.update.error ?? '-'}`,
    ``,
    `## Last error`,
    input.lastError ?? '-',
    ``,
    `## Log (${logPath() ?? 'file unavailable'})`,
    logText() || '(empty)',
  ]
  return lines.join('\n')
}

File: mobile/app-desktop/src/main/kotlin/com/silencelen/huginn/desktop/diag/Diagnostics.kt
package com.silencelen.huginn.desktop.diag

/**
 * The "Copy diagnostics" payload: everything worth pasting into a chat when
 * something looks wrong, and nothing that must not leave the machine.
 *
 * THE TOKEN IS EXCLUDED BY CONSTRUCTION. [Input] has no field it could travel in
 * — only `hasToken`, a boolean — so there is no call site that could put it there
 * by accident, and no future edit to this file that could start including it
 * without also changing the type. [RingLog.scrub] over the finished text is the
 * second line of defence, for the token that arrives inside somebody else's error
 * message ("401 for Bearer abc…").
 *
 * Pure: it takes facts and returns text. Gathering the facts is [collect], which
 * is where the runtime lives.
 */
object Diagnostics {

    /**
     * Everything the report can say. Note what is NOT here: the bearer token, the
     * settings file's contents, anything read off disk. Adding a field is adding a
     * thing the owner will paste into a chat window.
     */
    data class Input(
        val generatedAt: String,
        val appVersion: String,
        val packaged: Boolean,
        val platform: String,
        val jvm: String,
        val uptimeSec: Long,
        val heapUsedMb: Long,
        val heapMaxMb: Long,
        val baseUrl: String,
        val routePinned: Boolean,
        val hasToken: Boolean,
        val clientId: String,
        val watchConnected: Boolean,
        val lastWatchError: String?,
        val lastWatchErrorAt: String?,
        val appdVersion: String?,
        val notifyEnabled: Boolean,
        val present: Boolean,
        val visible: Boolean,
        val claiming: Boolean,
        /** The notification backend's name, or null when none is wired. */
        val notifier: String?,
        val updateStatus: String,
        val updateVersion: String?,
        val updateError: String?,
        val lastError: String?,
        val logPath: String?,
        val log: String,
    )

    fun build(input: Input): String {
        val lines = buildList {
            add("# Huginn Desktop diagnostics")
            add("generated       ${input.generatedAt}")
            add("")
            add("## App")
            add("version         ${input.appVersion}")
            add("packaged        ${input.packaged}")
            add("platform        ${input.platform}")
            add("jvm             ${input.jvm}")
            add("uptime          ${input.uptimeSec}s")
            add("heap            ${input.heapUsedMb}MB used / ${input.heapMaxMb}MB max")
            add("")
            add("## Connection")
            add("server          ${input.baseUrl}${if (input.routePinned) " (pinned)" else ""}")
            // The one honest thing to say about a secret in a shareable report.
            add("token           ${if (input.hasToken) "set" else "MISSING"}")
            add("client id       ${input.clientId}")
            add("watch stream    ${if (input.watchConnected) "connected" else "DISCONNECTED"}")
            add("last watch err  ${dash(input.lastWatchError)}${input.lastWatchErrorAt?.let { " at $it" } ?: ""}")
            add("appd version    ${input.appdVersion ?: "unknown"}")
            add("")
            add("## Notifications")
            add("enabled         ${input.notifyEnabled}")
            add("window visible  ${input.visible}")
            add("attended        ${input.present}")
            add("claiming route  ${input.claiming}")
            add("desktop notifier ${input.notifier ?: "NOT WIRED (Telegram and the phone remain the only routes)"}")
            add("")
            add("## Update")
            add("status          ${input.updateStatus}")
            add("version         ${dash(input.updateVersion)}")
            add("error           ${dash(input.updateError)}")
            add("")
            add("## Last error")
            add(dash(input.lastError))
            add("")
            add("## Log (${input.logPath ?: "file unavailable"})")
            add(input.log.ifBlank { "(empty)" })
        }
        // Scrubbed as a WHOLE, after assembly: a credential that arrived inside a
        // free-text error message is still a credential, and the fields it can hide
        // in are exactly the ones no schema constrains.
        return RingLog.scrub(lines.joinToString("\n"))
    }

    private fun dash(s: String?): String = if (s.isNullOrBlank()) "-" else s
}
